src/rust/ensogl/src/display/shape/text/text_field.rs

```diff
@@ -77,6 +150,16 @@ shared! { TextField
 
         /// Scroll text by given offset in pixels.
         pub fn scroll(&mut self, offset:Vector2<f32>) {
+            let scroll_position = self.scroll_position();
+            let offset_y        = offset.y.min(scroll_position.y);
+            let padding_lines   = 2;
+            let lines           = self.content.lines().len() + padding_lines;
+            let text_height     = self.content.line_height * lines as f32;
+            let view_height     = self.size().y;
+            let height          = (text_height - view_height).max(0.0);
+            let offset_y        = offset_y.max(scroll_position.y - height);
+            let offset          = Vector2::new(offset.x, offset_y);
+
```

Comment:
```suggestion
            let scroll_position = self.scroll_position();
            let padding_lines   = 2;
            let lines           = self.content.lines().len() + padding_lines;
            let text_height     = self.content.line_height * lines as f32;
            let view_height     = self.size().y;
            let height          = (text_height - view_height).max(0.0);
            let offset_y        = offset.y.min(scroll_position.y).max(scroll_position.y - height);
            let offset          = Vector2::new(offset.x, offset_y);
```
A little bit clearer for me.